Change API to make possible serialize/deserialize duktape function

mosolovsa opened the pull request:
1. Changed API with respect of the rule when duktape C function returns (void *), API returns (unsafe.Pointer).
2. Added example with serializtion/deserialization duktape function to/from bytecode.

Commit:
Change API to make possible serilize/deseiralize duktape function

1. Changed API with respect of the rule when duktape C function returns (void *), API returns (unsafe.Pointer)
2. Added example with serializtion/deserialization duktape function to/from bytecode

Commit:
Update main.go

olebedev reviewed (changes requested):
Hi  @mosolovsa, thanks for the contribution.
May I ask you to add _godoc_ compatible example instead?

It seems it would be more useful.

mosolovsa commented:
Hi @olebedev, thanks for the quick reply.
I will add godoc compatible example instead of source code example.

Commit:
Added godoc compatiable example

Commit:
removed examples folder

Commit:
Bind example to Context and run gofmt

mosolovsa commented:
I've removed folder examples from the feature branch and added file `example_bytecode_test.go` to implement godoc compatible example.
I'm new to github and golang and I'm not sure that this is exactly what you want. If it's not please kindly explain what exactly should I change.

olebedev commented on example_bytecode_test.go:
```diff
@@ -0,0 +1,107 @@
+	"reflect"
+	"unsafe"
+
+	"gopkg.in/olebedev/go-duktape.v2"
```
`v3` ?

olebedev commented on example_bytecode_test.go:
```diff
@@ -0,0 +1,107 @@
+	"gopkg.in/olebedev/go-duktape.v2"
+)
+
+func ExampleContext() {
```
It seems we could give a more relevant name for this example: `func ExampleSerializeFunction() {`

olebedev commented:
@mosolovsa, I've added a couple of comments to the changes, take a look.

@pytlesk4, will you check?

Commit:
Channged example name from ExampleContext to ExampleSerializeFunction

1.Channged example name from ExampleContext to ExampleSerializeFunction
2. Use go-duktape.v3 import in ExampleSerializeFunction

mosolovsa commented:
@olebedev Thanks for the review. I've changed example name from ExampleContext to ExampleSerializeFunction.
Looks like this is not how the godoc naming convention works. After that change added example missing from godoc.
https://godoc.org/github.com/mosolovsa/go-duktape
More details explained at https://blog.golang.org/examples in the "Larger example" section.
In short, as I understood you bind your example to specific section of godoc by example function name.
You can bind your example to specific type by naming example function as "Example{typename}", or bind to specific method of a type by "Example{typename}_{typemethodname}".
Another possibility is to create package level Example by naming example function as just "Example".
Or we can divide our large example to serialize examprle and bind it to Context.DumpFunction and deserialize example and bind it to Context.LoadFunction.

olebedev commented:
@mosolovsa, ufff, didn't know that, thank you! Ask you to make it visible and give a reasonable name.

Commit:
Bind serialization example to "func (*Context) LoadFunction" section

Commit:
okay, now I've really done this

mosolovsa commented:
I don't want to split this example to two separate example because it seems to me that idiomatic godoc example must be kind of a test with output.
Also, I don't see any possibility to specify example name, you can bind example to specific section of godoc (bind it to specific type or a type method, or make example as a package level example).
I've binded example to "func (*Context) LoadFunction" section. Please confirm if that's appropriate way.

http://godoc.org/github.com/mosolovsa/go-duktape#example-Context-LoadFunction
Now this section has an example.

Commit:
make example applicable as a test

olebedev commented:
@mosolovsa, please take a look - https://app.wercker.com/olebedev/go-duktape/runs/build/59802c5fc3a2eb000144ed60?step=59802c65d0736a0001f763de

It seems we have to use `package duktape` instead of `package duktape_test` in the file `./example_bytecode_test.go`

Commit:
Change package of the example from duktape_test to duktape

mosolovsa commented:
@olebedev I've changed `package duktape_test` to `package duktape`.

I'm not an expert of Wercker CI, but it looks like setup script should be modified to get dependencies for test (e.g. missing package `gopkg.in/olebedev/go-duktape.v3` on the Wercker CI host environment).

Here's an article describing something similar http://blog.wercker.com/publishing-go-binaries-with-wercker in the section `Building the application`.

However I don't know what exactly changes should be done, so may be moving example function away from test package is a better solution for now. However, automated test for example function wouldn't run.

EDIT: I've used `_test` suffix in package from example file as it is in https://blog.golang.org/examples. Looks like it's not necessary.

olebedev commented:
@mosolovsa, all you need is just drop import of `gopkg.in/olebedev/go-duktape.v3` and remove references to `duktape` package, like this:
```diff
--	ctxSerialize := duktape.New()
++	ctxSerialize := New()
```

Just make test passing locally.

Commit:
remove duktape from imports in package duktape

mosolovsa commented:
@olebedev Yeah, I should noticed that, thanks for the correction. Done.

pytlesk4 commented on example_bytecode_test.go:
```diff
@@ -1,4 +1,4 @@
-package duktape_test
+package duktape
```
Can we change the package back to duktape_test? Keeping example code in a different package forces you to write like a user.

mosolovsa replied:
I don't mind, but then you need to setup Wercker CI initialize script to download dependencies (e.g. to execute command "go get gopkg.in/olebedev/go-duktape.v3") or it would be broken.

olebedev replied:
@pytlesk4, in this case, we won't be able to test internals as we do it now. Also, we will have to rewrite all the tests as a separate package. Are you sure that it's good? I am not.

mosolovsa replied:
Another opportunity is to move example to README.md. As I can see there is already couple of examples thou.

pytlesk4 replied:
@olebedev we can leave the test we have already as is, we don't need to change that. Just the example, I can make the change right now. I don't thinks we need to do anything with the readme.

pytlesk4 replied:
https://app.wercker.com/olebedev/go-duktape/runs/build/5981f5633b92d9000165b177

pytlesk4 reviewed (commented):
Revert please

pytlesk4 commented on example_bytecode_test.go:
```diff
@@ -5,8 +5,6 @@ import (
 	"unsafe"
 	"log"
-
-	"gopkg.in/olebedev/go-duktape.v3"
```
import duktape this way: github.com/olebedev/go-duktape

olebedev commented:
@pytlesk4, there is nothing to revert, the PR is not merged yet. What do you mean?

Commit:
Change API in case of C pointer usage

To not disguise the fact that int can change due to function execution

mosolovsa commented:
My bad, I forgot to change API to use pointer to int in all cases duktape C API wants a pointer to int. Fixed. I think it would be better so user of the duktape Go API passing pointer and expecting the value can change after execute of a function.

olebedev reviewed (dismissed):
~

Commit:
Change API in case of C pointer usage

Changed in all occurencies of int as pointer, not only the functions that I've changed earlier.

pytlesk4 commented on api.go:
```diff
@@ -492,8 +492,8 @@ func (d *Context) GetLength(index int) int {
 // See: http://duktape.org/api.html#duk_get_lstring
-func (d *Context) GetLstring(index int, outLen int) string {
-	if s := C.duk_get_lstring(d.duk_context, C.duk_idx_t(index), (*C.duk_size_t)(unsafe.Pointer(&outLen))); s != nil {
+func (d *Context) GetLstring(index int, outLen *int) string {
```
Did we have to change outLen to a pointer? The way we were doing this before looks fine. If so, let's create a v4 branch and bump the version. That way anyone using v3 won't break because of the api changing.

mosolovsa replied:
https://play.golang.org/p/A-LOLTiFu0

mosolovsa replied:
Without pointer you can't get the buffer outLen outside of a GetLstring function. E.g. in the example that I provided I need to get the length of a buffet with bytecode. I'm not sure my English is good enough to explain what I mean so I provided link to code earlier.

pytlesk4 replied:
I understand what you are saying, your english is totally fine. I think it makes sense to update the api now, especially after looking at duktapes docs again. What about updating the apis to something like below:

```go
func (d *Context) GetLstring(index int, outLen int) (string, int) {
  if s := C.duk_get_lstring(d.duk_context, C.duk_idx_t(index), (*C.duk_size_t)(unsafe.Pointer(&outLen))); s != nil {
    return C.GoString(s), outLen
  }
  
  return "", 0
}
```

This feels more like idiomatic go to me, and it makes the code a little bit cleaner, and easier for go-duktape users to update their code. For example, [example_byte_code.go](https://github.com/mosolovsa/go-duktape/blob/BytecodeSerialization/example_bytecode_test.go#L37-L43) could be:
```go
// void *duk_get_buffer(duk_context *ctx, duk_idx_t idx, duk_size_t *out_size);
// typedef size_t duk_size_t;
// duk_size_t equivalent in Go is uint, however slice len is of a type int
// and I don't think that the limit of max(int) can be reached so for the sake of simplicity
// lets assume that bytecode size is always can be contained in int
rawmem, sz := ctxSerialize.GetBuffer(-1, 0)
```
Thoughts? @mosolovsa @olebedev Sorry about all the back and forth, but depending on what you guys think this should be the last issue. Thanks for the pull request!

mosolovsa replied:
```go
func (d *Context) GetLstring(index int) (string, int) {
  var outLen int
  if s := C.duk_get_lstring(d.duk_context, C.duk_idx_t(index), (*C.duk_size_t)(unsafe.Pointer(&outLen))); s != nil {
    return C.GoString(s), outLen
  }
  
  return "", 0
}
```

Isn't it would be better, you don't need to provide outLen because it would be obtained by API user as a return value.

Also if you're thinking about new API please mention that duktape C API returning outLen of type uint. So, I think there may be troubles with int overflow after value 0x7FFFFFFF on 32 bit arch., e.g. 0x80000000 interpreted as int would be -2147483648, but as uint it's a valid outLen 2147483648. 0x80000001 interpreted as int would be -2147483647, but as uint it's a valid outLen 2147483649 and so on.
However, I'm pretty sure this is not the real case.

pytlesk4 replied:
Nice, I like it!

mosolovsa replied:
> @mosolovsa, please take a look - https://app.wercker.com/olebedev/go-duktape/runs/build/59802c5fc3a2eb000144ed60?step=59802c65d0736a0001f763de

Does this issue solved?

pytlesk4 replied:
Strange, it looks like you might need to pull down the latest changes: https://github.com/mosolovsa/go-duktape/commit/318b65661207e8d88941cee476518009e39ab321

That commit fixed that issue.

olebedev replied:
@pytlesk4, at the beginning this lib was started as a transparent shim between C and Go. Due to that fact that in Go some things could be done in a more elegant way, I would say that we can accept some trade-offs for more convenience. 

In this certain case, I would return the value instead of passing a pointer b/c it's more safety. Also, do we need to return a length of the buffer if we have the buffer? It seems no.

Also, I do not agree with you about test package name. It's very much ok if we will have some examples right at the godoc page. If we split test packages(which is non-idiomatic at all) we will lose this ability.

@mosolovsa ask you to revert test package to `duktape`(at it was suggested previously) and fix buffers/strings lengths. I think we shouldn't return it b/c we have the value which could be measured. 

Correct my guys if you think I am wrong. Open to discussion.

olebedev replied:
@pytlesk4, besides, we don't have to increase the version b/c previous implementation was incorrect. So, it's just a butfix.

mosolovsa replied:
> Also, do we need to return a length of the buffer if we have the buffer?

In my implementation we return a unsafe.Pointer, not a buffer, to give the user more flexibility and let decide what to do with this raw pointer. unsafe.Pointer doesn't have length.

> I think we shouldn't return it b/c we have the value which could be measured.

I don't understand what exactly do you mean. If you mean that length contained in byte slice than we don't return the byte slice now, just a raw pointer, just as C.
Should we return byte slice? And force user to use byte slice (It seems ok to me because go is a high level language and as far as I can see std go library functions returning unsafe.Pointer not too often)? Or should we map C API 1 on 1 and pass a int pointer to obtain buffer length? Or get buffer length as a return value from function?

> @pytlesk4, besides, we don't have to increase the version b/c previous implementation was incorrect. So, it's just a butfix.

Yeah, it seems to me as a bugfix and I think it should be merged in all versions, but it's up to you.

EDIT: If we force user to use byte slice than we have allocation on heap because slice must be created inside of a function. Allocations on heap increase load on garbage collector (more references to check for free), that's not a appropriate behavior in some cases.
If we wouldn't create the new slice inside of a function and create slice header to the unsafe.Pointer, than the memory would be freed on context.HeapDestroy() and using that slice would lead to segfault which is very bad.
I think it would be better to return unsafe.Pointer with buffer length to let user decide what to do with it.


Or provide two versions:

1. return byte slice

2.  return unsafe.Pointer and buffer length

pytlesk4 commented on api.go:
```diff
@@ -1166,8 +1166,8 @@ func (d *Context) RequireBoolean(index int) bool {
 // See: http://duktape.org/api.html#duk_require_buffer
-func (d *Context) RequireBuffer(index int, outSize int) unsafe.Pointer {
-	return C.duk_require_buffer(d.duk_context, C.duk_idx_t(index), (*C.duk_size_t)(unsafe.Pointer(&outSize)))
+func (d *Context) RequireBuffer(index int, outSize *int) unsafe.Pointer {
```
Do we really need to pass an int pointer?

pytlesk4 commented on api.go:
```diff
@@ -1198,8 +1198,8 @@ func (d *Context) RequireInt(index int) int {
 // See: http://duktape.org/api.html#duk_require_lstring
-func (d *Context) RequireLstring(index int, outLen int) string {
-	if s := C.duk_require_lstring(d.duk_context, C.duk_idx_t(index), (*C.duk_size_t)(unsafe.Pointer(&outLen))); s != nil {
+func (d *Context) RequireLstring(index int, outLen *int) string {
```
Do we really need to pass an int pointer?

pytlesk4 commented on api.go:
```diff
@@ -1290,8 +1290,8 @@ func (d *Context) SafeCall(fn, args *[0]byte, nargs, nrets int) int {
 // See: http://duktape.org/api.html#duk_safe_to_lstring
-func (d *Context) SafeToLstring(index int, outLen int) string {
-	if s := C.duk_safe_to_lstring(d.duk_context, C.duk_idx_t(index), (*C.duk_size_t)(unsafe.Pointer(&outLen))); s != nil {
+func (d *Context) SafeToLstring(index int, outLen *int) string {
```
Last one, in general, totally cool with changing the api, but we need to version it for breaking changes. So if we don't need to pass an int pointer let's change it back to an int please.

pytlesk4 committed:
Put example in new package

This should fix the build error.

Commit:
Redesigned API


Redesigned API

mosolovsa commented:
@olebedev @pytlesk4 Please check if my last commit is acceptable.

pytlesk4 commented:
Looks good to me!

pytlesk4 reviewed (approved):
Thank you very much. This is awesome.

olebedev commented:
Thanks @mosolovsa, good improvement

mosolovsa commented:
Thanks for collaboration, guys!
